src/java.base/share/classes/java/lang/System.java

```diff
@@ -2113,6 +2118,13 @@ private static void initPhase1() {
         VM.saveProperties(tempProps);
         props = createProperties(tempProps);
 
+        // Force the jnu encoding to UTF-8, if it is not supported
+        jnuEncoding = props.getProperty("sun.jnu.encoding");
+        jnuEncodingSupported = jnuEncoding != null && Charset.isSupported(jnuEncoding);
+        if (!jnuEncodingSupported) {
+            props.setProperty("sun.jnu.encoding", "UTF-8");
+        }
+
```

Comment:
You could replace the two fields with something like "notSupportedJnuEncoding" that is only set when not supported. That keeps the additional code in initPhase1 to a minimum.

Reply:
Replaced as suggested.